Refactor descriptor set usage

zakorgy opened the pull request:
This is the first step to make the descriptor creation more transparent. Also this makes the pipeline layout creation more straightforward.
Now that we have well defined `DescriptorSetLayout`'s, as a follow up we can start use `rendy-descriptor`, and drop the `descriptor` module.
This PR adds some validation layer errors because it's in a transition state: for pipeline_layouts we use `DescriptorSetLayout`s from the code, but for descriptor sets we still use the ones from the `ron` file which is generated by the build script. After we move to `rendy-descriptor` this will be fixed.

zakorgy commented on webrender/src/device/gfx/descriptor_set.rs:
```diff
@@ -0,0 +1,167 @@
+
+pub(super) const DEFAULT_SET_1: &'static [DescriptorSetLayoutBinding] = &[
+    // Color0 sampler
+    descriptor_set_layout_binding!(12, DT::Sampler, SSF::ALL, false),
```
I should check why these numbers start at `12` :suspect:

kvark replied:
yeah, there is no need to use anything but the regular 0, 1, 2, 3, etc, since the namespaces of different descriptor set layouts are independent

kvark commented on webrender/src/device/gfx/descriptor_set.rs:
```diff
@@ -0,0 +1,167 @@
+use hal::pso::{DescriptorSetLayoutBinding, DescriptorType as DT, ShaderStageFlags as SSF};
+use super::super::{ShaderKind, VertexArrayKind};
+
+macro_rules! descriptor_set_layout_binding {
```
Let's replace this macro with a simple `const fn`

kvark commented on webrender/src/device/gfx/descriptor_set.rs:
```diff
@@ -0,0 +1,167 @@
+
+pub(super) const DEFAULT_SET_0: &'static [DescriptorSetLayoutBinding] = &[
+    // Dither
+    descriptor_set_layout_binding!(0, DT::SampledImage, SSF::ALL, false),
```
we should clean up the `SSF` use here and only select appropriate stages

kvark commented on webrender/src/device/gfx/descriptor_set.rs:
```diff
@@ -0,0 +1,167 @@
+];
+
+pub(super) const DEFAULT_SET_2: &'static [DescriptorSetLayoutBinding] = &[
+    // Color0
```
why are we providing color textures separately from color samplers? If they are meant to be together all the time, we should just go for `CombinedImageSampler`. Or do we know the samplers earlier (and they change less often) than the images?

kvark commented on webrender/src/device/gfx/descriptor_set.rs:
```diff
@@ -0,0 +1,167 @@
+    descriptor_set_layout_binding!(13, DT::Sampler, SSF::ALL, false),
+    // Color2 sampler
+    descriptor_set_layout_binding!(14, DT::Sampler, SSF::ALL, false),
+    // Dither sampler
```
dither sampler should go into SET 0

kvark commented on webrender/src/device/gfx/descriptor_set.rs:
```diff
@@ -0,0 +1,167 @@
+    descriptor_set_layout_binding!(1, DT::Sampler, SSF::ALL, false),
+    // GpuCache sampler
+    descriptor_set_layout_binding!(2, DT::Sampler, SSF::ALL, false),
+    // TransformPalette sampler
```
these samplers for global things should also go to SET 0

kvark commented on webrender/src/device/gfx/descriptor_set.rs:
```diff
@@ -0,0 +1,167 @@
+    descriptor_set_layout_binding!(13, DT::Sampler, SSF::ALL, false),
+    // Color2 sampler
+    descriptor_set_layout_binding!(14, DT::Sampler, SSF::ALL, false),
+    // PrevPassAlpha sampler
```
Samplers of per-pass data should go to SET 0. Pretty sure you can get away with a single sampler to be used for both color and alpha, no need for 2.

kvark commented on webrender/src/device/gfx/descriptor_set.rs:
```diff
@@ -0,0 +1,167 @@
+    descriptor_set_layout_binding!(15, DT::Sampler, SSF::ALL, false),
+    // PrevPassColor sampler
+    descriptor_set_layout_binding!(16, DT::Sampler, SSF::ALL, false),
+    // Dither sampler
```
similarly, this one can go to SET 0, as well as other global samplers

kvark commented on webrender/src/device/gfx/descriptor_set.rs:
```diff
@@ -0,0 +1,167 @@
+    // Color2
+    descriptor_set_layout_binding!(2, DT::SampledImage, SSF::ALL, false),
+    // PrevPassAlpha
+    descriptor_set_layout_binding!(3, DT::SampledImage, SSF::ALL, false),
```
these could go into a separate descriptor set, given that they only change per render target

kvark commented on webrender/src/device/gfx/device.rs:
```diff
@@ -240,6 +241,7 @@ pub struct Device<B: hal::Backend> {
     bound_desc_set_resources: DescriptorSetResources,
     descriptor_pools_per_frame: SmallVec<[DescriptorPools<B>; 1]>,
     descriptor_pools_sampler: SmallVec<[DescriptorPools<B>; 1]>,
+    descriptor_set_layouts: FastHashMap<DescriptorGroup, SmallVec<[B::DescriptorSetLayout; 3]>>,
```
this should use `ArrayVec` instead, given that we know the limit

kvark commented on webrender/src/device/gfx/device.rs:
```diff
@@ -570,6 +572,51 @@ impl<B: hal::Backend> Device<B> {
+        let descriptor_set_layouts = [DescriptorGroup::Default, DescriptorGroup::Clip, DescriptorGroup::Primitive]
+            .iter()
+            .map(|g| {
+                let layouts = match g {
```
let's do `layout_bindings = match g { .. }` returning the static slice of set layouts, so that we can have a single `create_descriptor_set_layout` invocation afterwards

kvark commented on webrender/src/device/gfx/device.rs:
```diff
@@ -1356,14 +1402,11 @@ impl<B: hal::Backend> Device<B> {
-        let shader_group = ShaderGroup::from(*shader_kind);
-        if let Entry::Vacant(v) = self.pipeline_layouts.entry(*shader_kind) {
+        let desc_group = DescriptorGroup::from(*shader_kind);
+        if let Entry::Vacant(v) = self.pipeline_layouts.entry(desc_group) {
```
why don't we create all these layouts at start?

zakorgy commented:
Added combined image samplers and immutbale samplers.
Added rendy-descriptor, and applied the descriptor reuse logic to all resource group.

kvark commented on webrender/src/device/gfx/device.rs:
```diff
@@ -1492,43 +1648,155 @@ impl<B: hal::Backend> Device<B> {
+                        src_stage.as_mut(),
+                    ) {
+                        unsafe {
+                            cmd_buffer.pipeline_barrier(
```
we should collect all barriers first and then issue at once, instead of issuing one by one

kvark commented on webrender/src/device/gfx/device.rs:
```diff
@@ -1492,43 +1648,155 @@ impl<B: hal::Backend> Device<B> {
+                    let mut src_stage = Some(hal::pso::PipelineStage::empty());
+                    if let Some(barrier) = image.transit(
+                        hal::image::Access::SHADER_READ,
+                        hal::image::Layout::ShaderReadOnlyOptimal,
```
I wonder if there needs to be any transition here at all. I'd assume that for most resources that are part of descriptor sets being `ShaderReadOnlyOptimal` is a normal state to be in. If, say, an image needs to receive some buffer transfers, it would change the layout, accept the transfer, and then change back the layout to `ShaderReadOnlyOptimal`. So by the time we reach here, we just need to assert that an image is in the proper state.

kvark commented on webrender/src/device/gfx/device.rs:
```diff
@@ -2607,17 +2903,31 @@ impl<B: hal::Backend> Device<B> {
+            self.per_draw_descriptor_sets.push(desc_set.unwrap());
+        }
+
+        let keys_to_remove: Vec<_> = self.per_pass_descriptor_bindings
```
similarly, this piece of logic should be abstracted away and not repeated

kvark commented on webrender/src/device/gfx/program.rs:
```diff
@@ -434,6 +434,35 @@ impl<B: hal::Backend> Program<B> {
             }
+            return
+        }
+        if let Some(binding) = self.bindings_map.get(&("s".to_owned() + binding)) {
```
nit: just use `format!`

kvark commented on webrender/src/device/gfx/program.rs:
```diff
@@ -407,6 +406,7 @@ impl<B: hal::Backend> Program<B> {
         cmd_buffer: &mut hal::command::CommandBuffer<B, hal::Graphics>,
+        sampler: Option<&B::Sampler>,
     ) {
         if let Some(binding) = self.bindings_map.get(&("t".to_owned() + binding)) {
```
we shouldn't be heap allocating ("to_owned") here on every move. A we shouldn't even string compare here, it's a relatively hot spot.

kvark commented on webrender/src/device/gfx/device.rs:
```diff
@@ -670,52 +634,40 @@ impl<B: hal::Backend> Device<B> {
+            &mut desc_allocator,
+            &descriptor_set_layouts[&DescriptorGroup::Default][DESCRIPTOR_SET_SAMPLER],
+            descriptor_set_ranges[&DescriptorGroup::Default][DESCRIPTOR_SET_SAMPLER],
+            27, // 3 * 3 * 3 is the maximum possible combination of filtering for the three mutable samplers
```
this should be computed from constants

kvark commented on webrender/src/device/gfx/device.rs:
```diff
@@ -1817,58 +1767,32 @@ impl<B: hal::Backend> Device<B> {
-                    let image = &self.images[&self.bound_textures[index]].core;
+                let desc_set = v.insert(descriptor_sets.pop().unwrap());
+                for &(index, sampler_name) in SAMPLERS[range].iter() {
+                    program.bound_textures[index] = bound_textures[index];
```
why are we still tracking textures per program? This was required for GL, but it's not like the binding model works in Vulkan anyway

zakorgy commented on webrender/build_gfx.rs:
```diff
@@ -28,18 +27,12 @@ const VK_EXTENSIONS: &'static str = "#extension GL_ARB_shading_language_420pack
-const DESCRIPTOR_SET_COUNT: usize = 3;
-#[cfg(not(feature = "push_constants"))]
+const DESCRIPTOR_SET_PER_PASS: usize = 0;
+const DESCRIPTOR_SET_PER_FRAME: usize = 1;
```
I assigned the `0` index to this set, because it's only used in a subset of shaders, so I can skip it when I bind descriptor sets to the pipeline layout by setting the first descriptor set index to `1` if the pipeline layout doesn't contains this set: https://github.com/szeged/webrender/pull/291/files#diff-fa2ec06e095c83d43cdc6e509e90c761R547

kvark replied:
when you bind descriptor sets *within* a pass, you wouldn't need to touch any of these two anyway

zakorgy commented on webrender/src/device/gfx/device.rs:
```diff
@@ -713,36 +881,38 @@ impl<B: hal::Backend> Device<B> {
 
         self.render_pass.take().unwrap().deinit(&self.device);
 
-        for pools in self.descriptor_pools_per_draw.iter_mut() {
```
We started to address this in https://github.com/szeged/webrender/pull/294

zakorgy commented on webrender/src/device/gfx/device.rs:
```diff
@@ -567,7 +588,138 @@ impl<B: hal::Backend> Device<B> {
+                    .map(|(index, (bindings, immutable_samplers))| {
+                        let layout = unsafe { device.create_descriptor_set_layout(*bindings, immutable_samplers.iter().map(|s| *s)) }
+                            .expect("create_descriptor_set_layout failed");
+                        if index == DESCRIPTOR_SET_PER_FRAME {
```
Because per frame descriptor has different layout for each descriptor group.

kvark replied:
I think we are using "per frame" wrong then. Maybe rename it to "per group"? That would address my concern about the descriptor set index as well.

kvark commented on webrender/src/device/gfx/descriptor.rs:
```diff
@@ -2,409 +2,519 @@
+                | ShaderKind::DebugColor | ShaderKind::DebugFont
+                | ShaderKind::VectorStencil | ShaderKind::VectorCover => DescriptorGroup::Default,
+            ShaderKind::ClipCache => DescriptorGroup::Clip,
+            _ => DescriptorGroup::Primitive,
```
if possible, let's try being more explicit about what those cases are instead of putting an all-catch "_"

kvark commented on webrender/src/device/gfx/descriptor.rs:
```diff
@@ -2,409 +2,519 @@
+
+ impl PartialEq for Locals {
+    fn eq(&self, other: &Locals) -> bool {
+        self.transform_as_u32_slice() == other.transform_as_u32_slice() &&
```
we can compare `uTransform` directly

kvark commented on webrender/src/device/gfx/descriptor.rs:
```diff
@@ -2,409 +2,519 @@
+pub(super) struct PerGroupBindings(pub [TextureId; PER_GROUP_TEXTURE_COUNT]);
+
+#[derive(Clone, Copy, Eq, PartialEq, Hash, Debug)]
+pub(super) struct SamplerBindings(pub [TextureFilter; MUTABLE_SAMPLER_COUNT]);
```
why are these separate instead of just being in one of the descriptor sets?

zakorgy replied:
Fixed in https://github.com/szeged/webrender/pull/291/commits/b5a153ba2d2b12dfdd1a0fe3a3ce055eb5da903f

kvark commented on webrender/src/device/gfx/descriptor.rs:
```diff
@@ -2,409 +2,519 @@
+pub(super) struct DescriptorGroupData<B: hal::Backend> {
+    pub(super) descriptor_set_layouts: FastHashMap<DescriptorGroup, ArrayVec<[B::DescriptorSetLayout; DESCRIPTOR_SET_PER_LAYOUT]>>,
+    pub(super) descriptor_set_ranges: FastHashMap<DescriptorGroup, ArrayVec<[DescriptorRanges; DESCRIPTOR_SET_PER_LAYOUT]>>,
+    pub(super) pipeline_layouts: FastHashMap<DescriptorGroup, B::PipelineLayout>,
```
there shouldn't be a need to have all these 3 maps separate: each has a key of `DescriptorGroup`, so we should just have a single map with values being structures of all the things we need

kvark commented on webrender/src/device/gfx/descriptor.rs:
```diff
@@ -2,409 +2,519 @@
+}
+
+impl<B: hal::Backend> DescriptorGroupData<B> {
+    pub(super) fn descriptor_layout(&self, group: &DescriptorGroup, group_idx: usize) -> &B::DescriptorSetLayout {
```
correspondingly, you'd not need any of these accessor methods if you only had a single map

kvark commented on webrender/src/device/gfx/descriptor.rs:
```diff
@@ -2,409 +2,519 @@
-            self.descriptor_set_layout,
-        ));
-        self.descriptor_set.push(desc_set);
+pub(super) trait FreeSets<B: hal::Backend> {
```
for things like this, it would be good to have a piece of documentation explaining what the trait is for, then on each method explaining what the method is for, specifically

kvark commented on webrender/src/device/gfx/descriptor.rs:
```diff
@@ -2,409 +2,519 @@
+                        unsafe {
+                            cmd_buffer.pipeline_barrier(
+                                src_stage.unwrap()
+                                    .. hal::pso::PipelineStage::FRAGMENT_SHADER,
```
why are we ending the transition at `FRAGMENT_SHADER`? If the descriptor has, say, GPU cache texture, it needs to be accessible from VS

kvark commented on webrender/src/device/gfx/descriptor.rs:
```diff
@@ -2,409 +2,519 @@
+                        src_stage.as_mut(),
+                    ) {
+                        unsafe {
+                            cmd_buffer.pipeline_barrier(
```
It would be good to combine all the transitions into a single `pipeline_barrier` call.
Like with any "would be good" things, please either address it in the PR, or add a descriptive TODO comment here explaining what remains to be done.

kvark commented on webrender/src/device/gfx/descriptor.rs:
```diff
@@ -2,409 +2,519 @@
+                        src_stage.as_mut(),
+                    ) {
+                        unsafe {
+                            cmd_buffer.pipeline_barrier(
```
similarly, would be good to combine these barriers into a single one.
All in all, maybe we can move the barrier generation out of the `match` arms, since it should be similar for both?

kvark commented on webrender/src/device/gfx/descriptor.rs:
```diff
@@ -2,409 +2,519 @@
+                        unsafe {
+                            cmd_buffer.pipeline_barrier(
+                                src_stage.unwrap()
+                                    .. hal::pso::PipelineStage::FRAGMENT_SHADER,
```
same concern about the `FRAGMENT_SHADER` part

kvark commented on webrender/src/device/gfx/descriptor.rs:
```diff
@@ -2,409 +2,519 @@
-        self.marked_brush_pools.clear();
-        self.marked_cache_clip_pools.clear();
+            });
+            unsafe { device.write_descriptor_sets(descriptor_writes) };
```
something to keep in mind - we could probably prepare these descriptors ahead of time of recording

kvark commented on webrender/src/device/gfx/descriptor.rs:
```diff
@@ -2,409 +2,519 @@
+        locals_buffer: &mut UniformBufferHandler<B>,
+        heaps: &mut Heaps<B>,
+    ) {
+        if let Entry::Vacant(v) = self.descriptor_bindings.entry(bindings) {
```
this part is *almost* being repeated multiple times, with a few differences
can we unify it?

kvark commented on webrender/src/device/gfx/device.rs:
```diff
@@ -1492,90 +1453,105 @@ impl<B: hal::Backend> Device<B> {
+        );
+        self.bound_per_group_textures = per_group_bindings;
+
+        // Immutable samplers
```
immutable samplers should be a part of the descriptor set layout, never need to be bound specifically
or perhaps these is about some other form of immutability?

zakorgy replied:
No, it's a typo, this should be `// Mutable samplers`.

zakorgy commented on webrender/src/device/gfx/descriptor.rs:
```diff
@@ -2,409 +2,519 @@
+pub(super) const PER_GROUP_RANGE_PRIMITIVE: std::ops::Range<usize> = 5..11; // GpuCache, TransformPalette, RenderTasks, Dither, PrimitiveHeadersF, PrimitiveHeadersI
+
+#[cfg(feature = "push_constants")]
+pub(super) const DESCRIPTOR_SET_PER_LAYOUT: usize = 4;
```
Another wrong naming, this supposed to be the count of used descriptor sets, changed it to `DESCRIPTOR_SET_COUNT`

zakorgy commented:
@kvark I think I have addressed most of your comments, and added a few TODO where work needs to be done, but it's outside of the scope of the current PR. Eventually we ended up with one descriptor set less, I moved the samplers from the sampler descriptor set into the per draw set.
I still want to keep the Locals in a separate descriptor set, since it's optional, we only use it if we can't use push constants.

kvark commented:
I haven't gone through the new code in detail. Let me know if there is a particular spot to review. We don't want too much churn on a PR and should move forward, addressing the non-critical concerns in follow-ups.
